### ISSUE TITLE
SDKMAN as installation alternative!

### ISSUE THREAD

**Issue by eprozium** [technomancy/leiningen#2204]
Please add SDKMAN as installation alternative to Lein :

http://sdkman.io/

It's very good at handling different versions in parallel, and it has API to push releases http://sdkman.io/vendors.html automatically. 

Ant, Maven, Gradle, SBT and other are already supporting this, and it would be practical if Lein would  too.

Thanks.

**Comment by hyPiRion** [comment]
I'll happily support this if it's not too much work to do so. :)

I looked at the vendoring and candidate releases, but couldn't immediately find out how the zip files should be structured, so I'll have a look at that when I'm able to.

**Comment by marc0der** [comment]
@hyPiRion I'd like to help you get set up on this. The zip file has the typical format that SDKs usually come in: in your case it will be a zip containing a `bin` folder, which in turn will hold your script (I believe that Lein has a single bootstrap script). The script needs to be executable.

In terms of getting the API credentials, you can contact me directly by email and I'll get those across to you.

**Comment by cprice404** [comment]
@hyPiRion let me know if you need help with this, I might be able to find some time.

**Comment by hyPiRion** [comment]
@marc0der: Just sent you an email, let's do this.

@cprice404: This would likely be something that we have to do during releases, but I don't mind explaining/pairing the 2.7.1 release with you if you're interested :)

**Comment by cprice404** [comment]
@hyPiRion sure, I'd be happy to participate if it would be helpful to you!  I'll be offline the rest of today most likely, but you can send me an email, or, I'll be online (and in #leiningen) all day tomorrow (US time).

**Comment by hyPiRion** [comment]
(Oops, wrong issue)

**Comment by marc0der** [comment]
The API credentials have been sent and the Candidate created. Now all that
remains is the API calls for the first version to be released.

**Comment by hyPiRion** [comment]
I'll try to look at this and get this done by today.

**Comment by hyPiRion** [comment]
This is now resolved, one can do `sdk install leiningen`, which will fetch Leiningen 2.7.1.

**Comment by eprozium** [comment]
Thank you very much!

Executing ```sdk list``` will list ```Leiningen (0.0.0)``` as version.

It seems to work however on Linux machines without problems.

On OSX (El Captain) it's not working :( : 
 - ```sdk install leiningen``` will show "Stop! leiningen is not a valid candidate." 
 - ```sdk list leiningen``` will show the same message
 - ```sdk list``` will list however ```Leiningen (0.0.0)``` as it does on Linux.

Would be much simpler if the candidate name would be ```lein``` :).

**Comment by hyPiRion** [comment]
Yeah, it seems like something interesting has happened with the release. I'll talk to the sdkman-people about it. 

For now, it _seems like_ it will work if you do `sdk selfupdate`, then reopen the terminal and type in `sdk install leiningen` (at least that was the case for me).

I think `sdk install leiningen` is alright. I'm not too worried about aliases, but if it's an option then that would be great. I'll bring it upstream, but if we're the only program with a long name and alias then I'd guess that this won't happen.

**Comment by marc0der** [comment]
I had a quick look and all is in order. The `sdk list` view lags behind a bit sometimes but is now up to date.

When a new candidate is introduced, you can simply do an `sdk flush candidates` and open a new terminal. I just took it for a spin and it works fine for me!

**Comment by hyPiRion** [comment]
Oh, excellent, thanks for verifying! :)

**Comment by marc0der** [comment]
@hyPiRion would you like it to be `lein` instead of `leiningen`? It would require that I regenerate your creds.

**Comment by hyPiRion** [comment]
@marc0der: It seems like the norm in sdk is the full name (e.g. `sdk install maven` instead of `sdk install mvn`), so there's no need to let Leiningen be an oddball here. I'm fine with the candidate name as it is, no worries.

**Comment by marc0der** [comment]
Great to have you on board!

**Comment by eprozium** [comment]
> For now, it seems like it will work if you do sdk selfupdate, then reopen the terminal and type in sdk install leiningen (at least that was the case for me).

Thank you ! This seems to do the trick on OSX.

>... would you like it to be lein instead of leiningen

I would. I can't count how often I mistyped ```leiningen``` . Not so with ```maven```.

**Comment by marc0der** [comment]
It had nothing to do with OSX, but rather that you already had sdkman
installed.

If the name is really a problem, I'll leave it up to the project lead to
decide.

**Comment by technomancy** [comment]
I would like to move the SDKMAN support into a separate repository for the downstream maintainer as per #2346; anyone here know who would be able to pick that up?

**Comment by hypirion** [comment]
I can probably pick this one up, as I was the one that brought it in. I'll put the stuff required in the https://github.com/orgs/leiningen org if that's fine with you (we already have chocolatey in there).

**Comment by rbjorklin** [comment]
There has been a few Leiningen releases made since this support was introduced. Is there anything I can do to assist in getting the latest Leiningen release available on SDKMAN?

**Comment by winks** [comment]
@hypirion if you could give me a quick rundown what you did, I can possibly move the sdkman stuff into a new repo in the leiningen org and update the release, I have some time over the weekend.

**Comment by tkaefer** [comment]
Do you have any updates here @hypirion ? SDKMAN still has version 2.7.1 - which seems a bit outdated.
It seems there are people to help here... 
Does it require a new issue, since this original one is closed?

**Comment by marc0der** [comment]
@tkaefer also something to be aware of is that community members can also add versions to sdkman by raising a PR here: https://github.com/sdkman/sdkman-db-migrations

**Comment by tkaefer** [comment]
Thanks for the hint @marc0der 👍
